File: backend/qms-system/src/main/java/nz/ac/autuni/qmssystem/model/RiskProcedure.java
package nz.ac.autuni.qmssystem.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import javax.persistence.GeneratedValue;
import java.io.Serializable;
import java.util.List;
import java.util.Map;

import static javax.persistence.GenerationType.IDENTITY;

/**
 * @author Simon-the-coder
 * @date 3/09/20 5:33 pm
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@Document(collection = "RiskProcedure")
public class RiskProcedure implements Serializable {

    @Id @GeneratedValue(strategy = IDENTITY)
    private String riskProcedureId;
    private String harm;
    private List<Item> severity;
    private List<ItemDescription> severityDescription;
    private List<Item> probability;
    private List<ItemDescription> probabilityDescription;
    private Map<String,String> mapRiskEstimation;
    private String approve;
}

@AllArgsConstructor
@NoArgsConstructor
@Data
class Item implements Serializable {
    private String id;
    private String name;
    private String level;
    private String tag;
}

@AllArgsConstructor
@NoArgsConstructor
@Data
class ItemDescription implements Serializable {
    private String itemId;
    private String name;
    private String description;
    private String tag;
}

File: frontend/QMS_system/lib/model/risk_procedure.dart
class RiskProcedure{
  String riskProcedureId;
  String harm;
  List<RiskProcedureItem> severity;
  List<RiskProcedureItemDescription> severityDescription;
  List<RiskProcedureItem> probability;
  List<RiskProcedureItemDescription> probabilityDescription;
  Map<String,dynamic> mapRiskEstimation;
  String approve;

  RiskProcedure({this.riskProcedureId, this.severity, this.severityDescription, this.probability, this.probabilityDescription, this.approve, this.harm, this.mapRiskEstimation});

  RiskProcedure.fromJson(Map<String, dynamic> json)
      : riskProcedureId = json['riskProcedureId'],
        harm = json['harm'],
        severity = json['severity'].map<RiskProcedureItem>((json) => RiskProcedureItem.fromJson(json)).toList(growable: false),
        severityDescription = json['severityDescription'].map<RiskProcedureItemDescription>((json) => RiskProcedureItemDescription.fromJson(json)).toList(growable: false),
        probability = json['probability'].map<RiskProcedureItem>((json) => RiskProcedureItem.fromJson(json)).toList(growable: false),
        probabilityDescription = json['probabilityDescription'].map<RiskProcedureItemDescription>((json) => RiskProcedureItemDescription.fromJson(json)).toList(growable: false),
        approve = json['approve'],
        mapRiskEstimation = json['mapRiskEstimation'];

  Map<String, dynamic> toJson() =>
      {
        'riskProcedureId': riskProcedureId,
        'harm': harm,
        'severity': severity.map<Map<String,dynamic>>((item) => item.toJson()).toList(growable: false),
        'severityDescription': severityDescription.map<Map<String,dynamic>>((item) => item.toJson()).toList(growable: false),
        'probability': probability.map<Map<String,dynamic>>((item) => item.toJson()).toList(growable: false),
        'probabilityDescription': probabilityDescription.map<Map<String,dynamic>>((item) => item.toJson()).toList(growable: false),
        'approve': approve,
        'mapRiskEstimation': mapRiskEstimation,
      };
}

class RiskProcedureItem {
  String id;
  String level;
  String name;
  String tag;

  RiskProcedureItem({this.id, this.level, this.name, this.tag});

  RiskProcedureItem.fromJson(Map<String, dynamic> json)
    : level = json['level'],
      name = json['name'],
      id = json["id"],
      tag = json["tag"];

  Map<String, dynamic> toJson() => {
    'level' : level,
    'name' : name,
    'id' : id,
    'tag' : tag
  };
}

class RiskProcedureItemDescription {
  String itemId;
  String name;
  String description;
  String tag;

  RiskProcedureItemDescription({this.itemId, this.description, this.name, this.tag});

  RiskProcedureItemDescription.fromJson(Map<String, dynamic> json)
      : description = json['description'],
        name = json['name'],
        itemId = json["itemId"],
        tag = json["tag"];

  Map<String, dynamic> toJson() => {
    'description' : description,
    'name' : name,
    'itemId' : itemId,
    'tag' : tag
  };
}